addons/html_builder/static/src/snippets/snippet_service.js

```diff
@@ -0,0 +1,417 @@
+import { AddSnippetDialog } from "./add_snippet_dialog";
+import { registry } from "@web/core/registry";
+import { user } from "@web/core/user";
+import { markup } from "@odoo/owl";
+import { RPCError } from "@web/core/network/rpc";
+
+export class SnippetModel extends Reactive {
+    constructor(services, { snippetsName, context }) {
+        super();
+        this.orm = services.orm;
+        this.dialog = services.dialog;
+        this.notification = services.notification;
+        this.snippetsName = snippetsName;
+        this.context = context;
+        this.loadProm = null;
+
+        this.snippetsByCategory = {
+            snippet_groups: [],
+            snippet_custom: [],
+            snippet_structure: [],
+            snippet_content: [],
+            snippet_custom_content: [],
+        };
+    }
+
+    get hasCustomGroup() {
+        return !!this.snippetsByCategory.snippet_custom.length;
+    }
+
+    get snippetGroups() {
+        const snippetGroups = this.snippetsByCategory.snippet_groups;
+        if (this.hasCustomGroup) {
+            return snippetGroups;
+        }
+        return snippetGroups.filter((snippet) => snippet.groupName !== "custom");
+    }
+
+    get snippetStructures() {
+        return [
+            ...this.snippetsByCategory.snippet_structure,
+            ...this.snippetsByCategory.snippet_custom,
+        ];
+    }
+
+    get snippetInnerContents() {
+        return this.snippetsByCategory.snippet_content;
+    }
+
+    get hasCustomInnerContents() {
+        return !!this.snippetsByCategory.snippet_custom_content.length;
+    }
+
+    get snippetCustomInnerContents() {
+        return this.snippetsByCategory.snippet_custom_content;
+    }
+
+    isCustomInnerContent(customSnippetName) {
+        return !!this.snippetsByCategory.snippet_content.find(
+            (snippet) => snippet.name === customSnippetName
+        );
+    }
+
+    isCustomStructure(customSnippetName) {
+        return !!this.snippetsByCategory.snippet_structure.find(
+            (snippet) => snippet.name === customSnippetName
+        );
+    }
+
+    getSnippet(category, id) {
+        return this.snippetsByCategory[category].find((snippet) => snippet.id === id);
+    }
+
+    getSnippetByName(category, name) {
+        return this.snippetsByCategory[category].find((snippet) => snippet.name === name);
+    }
+
+    installSnippetModule(snippet) {
+        // TODO: Should be the app name, not the snippet name ... Maybe both ?
+        const bodyText = _t("Do you want to install %s App?", snippet.title);
+        const linkText = _t("More info about this app.");
+        const linkUrl =
+            "/odoo/action-base.open_module_tree/" + encodeURIComponent(snippet.moduleId);
+
+        this.dialog.add(ConfirmationDialog, {
+            title: _t("Install %s", snippet.title),
+            body: markup(
+                `${escape(bodyText)}\n<a href="${linkUrl}" target="_blank">${escape(linkText)}</a>`
+            ),
```

Comment:
This one can be replaced by markup template which no longer flags the ci since today, is safe, easier to type, and escapes param by default.
```suggestion
            body: markup`${bodyText}\n<a href="${linkUrl}" target="_blank">${linkText}</a>`,
```